test/node/session-provider-async.js

```diff
@@ -0,0 +1,75 @@
+'use strict';
+
+const test = require('tap');
+
+const jstp = require('../..');
+
+const APP_NAME = 'APP_NAME';
+const TOKEN = 'TOKEN';
+
+const interfaces = {
+  iface: {
+    first(connection, token, callback) {
+      connection.session.set(TOKEN, token);
+      callback(null);
+    },
+    second(connection, callback) {
+      callback(null, connection.session.get(TOKEN));
+    },
+  },
+};
+
+const application = new jstp.Application(APP_NAME, interfaces);
+
+const TIMEOUT = 100;
+const map = new Map();
+const sessionStorageProvider = {
+  set: (sessionId, session) => {
```

Comment:
Now that I've seen this, I thought about this async set. Ain't there a race condition between `sessionStorageProvider.set` and user reconnect. So if server uses async sessionStorageProvider and user reconnects too fast he may actually not get a session or he will get an outdated (because the session wasn't saved yet) session, wdyt @nechaido @belochub ?
One possible solution may be adding a callback argument to `set` and only `this._clientsBySessionId.delete(connection.session.id);` in the server if we know that it was stored (after callback call).